File: Hash Map & Set/Problem3 - 1657.cpp
#include <bits/stdc++.h>
using namespace std;

/* optimal approach - using hahsing/set */
bool closeStrings(string word1, string word2) {
    int n = word1.size(), m = word2.size();
    if (n != m) return false;
    vector<int> freq1(26, 0), freq2(26, 0);
    unordered_set<char> set1, set2;
    for (auto& it : word1) {
        freq1[it - 'a']++;
        set1.insert(it);
    }
    for (auto& it : word2) {
        freq2[it - 'a']++;
        set2.insert(it);
    }

    sort(freq1.begin(), freq1.end()); 
    sort(freq2.begin(), freq2.end()); 
    return freq1 == freq2 && set1.size() == set2.size(); 
    /* TC-O(n + m + (2 * 26 * Log(26))) SC-O(2*26) */
}

File: Hash Map & Set/Problem3 - 1657.java
import java.util.Arrays;
import java.util.HashSet;

public class Main {
    /* optimal approach - using hashing/set */
    static boolean closeStrings(String word1, String word2) {
        int n = word1.length(), m = word2.length();
        if (n != m) return false;
        
        int[] freq1 = new int[26];
        int[] freq2 = new int[26];
        HashSet<Character> set1 = new HashSet<>();
        HashSet<Character> set2 = new HashSet<>();
        
        for (char ch : word1.toCharArray()) {
            freq1[ch - 'a']++;
            set1.add(ch);
        }
        
        for (char ch : word2.toCharArray()) {
            freq2[ch - 'a']++;
            set2.add(ch);
        }

        Arrays.sort(freq1);
        Arrays.sort(freq2);
        
        return Arrays.equals(freq1, freq2) && set1.equals(set2);
        /* TC-O(n + m + (2 * 26 * Log(26))) SC-O(2*26) */
    }

    public static void main(String[] args) {
        String word1 = "abc";
        String word2 = "bca";
        System.out.println("Are word1 and word2 close strings? " + closeStrings(word1, word2));
    }
}
